src/EFCore.PG/Infrastructure/Internal/NpgsqlOptionsExtension.cs

```diff
@@ -19,45 +20,93 @@
 // TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+
 #endregion
 
+using System;
 using System.Collections.Generic;
 using System.Net.Security;
 using JetBrains.Annotations;
 using Microsoft.EntityFrameworkCore.Infrastructure;
 using Microsoft.Extensions.DependencyInjection;
 using Npgsql.EntityFrameworkCore.PostgreSQL.Utilities;
 
-// ReSharper disable once CheckNamespace
 namespace Npgsql.EntityFrameworkCore.PostgreSQL.Infrastructure.Internal
 {
     public class NpgsqlOptionsExtension : RelationalOptionsExtension
     {
+        /// <summary>
+        /// The collection of database plugins.
+        /// </summary>
+        [NotNull] readonly List<NpgsqlEntityFrameworkPlugin> _plugins;
+
+        /// <summary>
+        /// The name of the database for administrative operations.
+        /// </summary>
+        [NotNull]
         public string AdminDatabase { get; private set; }
-        public bool? ReverseNullOrdering { get; private set; }
-        public ProvideClientCertificatesCallback ProvideClientCertificatesCallback { get; private set; }
-        public RemoteCertificateValidationCallback RemoteCertificateValidationCallback { get; private set; }
 
+        /// <summary>
+        /// The backend version to target.
+        /// </summary>
+        [NotNull]
+        public Version PostgresVersion { get; private set; }
+
+        /// <summary>
+        /// The collection of database plugins.
+        /// </summary>
+        [NotNull]
+        [ItemNotNull]
         public IReadOnlyList<NpgsqlEntityFrameworkPlugin> Plugins => _plugins;
 
-        readonly List<NpgsqlEntityFrameworkPlugin> _plugins = new List<NpgsqlEntityFrameworkPlugin>();
+        /// <summary>
+        /// The specified <see cref="ProvideClientCertificatesCallback"/>.
+        /// </summary>
+        [CanBeNull]
+        public ProvideClientCertificatesCallback ProvideClientCertificatesCallback { get; private set; }
 
-        public NpgsqlOptionsExtension() {}
+        /// <summary>
+        /// The specified <see cref="RemoteCertificateValidationCallback"/>.
+        /// </summary>
+        [CanBeNull]
+        public RemoteCertificateValidationCallback RemoteCertificateValidationCallback { get; private set; }
 
-        // NB: When adding new options, make sure to update the copy ctor below.
+        /// <summary>
+        /// True if reverse null ordering is enabled; otherwise, false.
+        /// </summary>
+        public bool ReverseNullOrdering { get; private set; }
 
-        public NpgsqlOptionsExtension([NotNull] NpgsqlOptionsExtension copyFrom)
-            : base(copyFrom)
+        /// <summary>
+        /// Initializes an instance of <see cref="NpgsqlOptionsExtension"/> with the default settings.
+        /// </summary>
+        public NpgsqlOptionsExtension()
+        {
+            AdminDatabase = "postgres";
+            PostgresVersion = new Version(10, 4);
```

Comment:
I'm not sure it's a good idea to hardcode a default version like this:

1. It makes it impossible to distinguish between user-provided values and defaults - a distinction that may be important. With this change, if we ask the question "what is the current database compatibility version", the answer will be "10.4" even if the user hasn't specified anything.
2. This value won't evolve unless we make it evolve with each and every PostgreSQL release.
3. Finally, I'd rather have the default admin database (`postgres`) specified in NpgsqlRelationalConnection

It seems safer and cleaner to just leave the options null if the user hasn't said anything. For `PostgresVersion`, a null would mean "assume the latest supported PostgreSQL version", i.e. don't turn off any features for backwards compat.

On the other hand, for `ReverseNullOrdering` I think your change is good - a nullable bool doesn't really help anyone and just makes things more complex.

Reply:
That makes sense. I'm going to punt on making any changes to the admin database stuff (since it was here before this PR)... But I can take a look at changing it in a follow up PR.

Reply:
Actually, scratch that. I thought that was a default from before I started this PR, but the diff looks like I've added it.... Dropping the hard-coded admin default.